File: bms/go/geo/geo_r3_PreciseVector_MulByFloat64/geo_r3_PreciseVector_MulByFloat64.go
package test

import "math/big"

var ()

const (
	geo_r3_prec = big.MaxPrec
)

func (v geo_r3_PreciseVector) MulByFloat64(f float64) geo_r3_PreciseVector {
	return v.Mul(geo_r3_precFloat(f))
}

func geo_r3_precFloat(f float64) *big.Float {
	return new(big.Float).SetPrec(geo_r3_prec).SetFloat64(f)
}

func (v geo_r3_PreciseVector) Mul(f *big.Float) geo_r3_PreciseVector {
	return geo_r3_PreciseVector{
		X: geo_r3_precMul(v.X, f),
		Y: geo_r3_precMul(v.Y, f),
		Z: geo_r3_precMul(v.Z, f),
	}
}

func geo_r3_precMul(a, b *big.Float) *big.Float {
	return new(big.Float).SetPrec(geo_r3_prec).Mul(a, b)
}

type geo_r3_PreciseVector struct {
	X, Y, Z *big.Float
}

File: translations_only/fail/go/geo/geo_r3_PreciseVector_MulByFloat64/claude3/Restart/geo_r3_PreciseVector_MulByFloat64.rs

use std::collections::VecDeque;

const GEO_R3_PREC: u64 = 1024;

fn geo_r3_precise_vector_mul_by_float64(v: geo_r3_PreciseVector, f: f64) -> geo_r3_PreciseVector {
    let f_prec = geo_r3_prec_float(f);
    geo_r3_precise_vector_mul(v, f_prec)
}

fn geo_r3_prec_float(f: f64) -> Box<BigInt> {
    let mut x = Box::new(BigInt::zero());
    x.set_prec(GEO_R3_PREC);
    x.set_float(f);
    x
}

fn geo_r3_precise_vector_mul(v: geo_r3_PreciseVector, f: Box<BigInt>) -> geo_r3_PreciseVector {
    geo_r3_PreciseVector {
        x: geo_r3_prec_mul(v.x, &f),
        y: geo_r3_prec_mul(v.y, &f),
        z: geo_r3_prec_mul(v.z, &f),
    }
}

fn geo_r3_prec_mul(a: Box<BigInt>, b: &BigInt) -> Box<BigInt> {
    let mut x = Box::new(BigInt::zero());
    x.set_prec(GEO_R3_PREC);
    x.mul(&a, b);
    x
}

struct geo_r3_PreciseVector {
    x: Box<BigInt>,
    y: Box<BigInt>,
    z: Box<BigInt>,
}

struct BigInt {
    data: VecDeque<u32>,
    prec: u64,
}

impl BigInt {
    fn zero() -> BigInt {
        BigInt { data: VecDeque::new(), prec: 0 }
    }

    fn set_prec(&mut self, prec: u64) {
        self.prec = prec;
    }

    fn set_float(&mut self, f: f64) {
        // Implementation omitted for brevity
    }

    fn mul(&mut self, a: &BigInt, b: &BigInt) {
        // Implementation omitted for brevity
    }
}
